From 28fc560016e79803bd0c0ffa8acd0d2f6bf8a0c9 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Mon, 11 Mar 2019 08:07:38 +0100
Subject: [PATCH 31/43] ethtool: set pause parameters with PARAMS_SET request

Add support for setting device pause parameters using SET_SETTINGS request
with ETHTOOL_A_PARAMS_PAUSE nested attribute. This is a replacement of
ETHTOOL_SPAUSEPARAM ioctl command.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 Documentation/networking/ethtool-netlink.txt | 14 +++--
 net/ethtool/params.c                         | 55 +++++++++++++++++++-
 2 files changed, 63 insertions(+), 6 deletions(-)

diff --git a/Documentation/networking/ethtool-netlink.txt b/Documentation/networking/ethtool-netlink.txt
index 707271ac61d7..dc0ad35944aa 100644
--- a/Documentation/networking/ethtool-netlink.txt
+++ b/Documentation/networking/ethtool-netlink.txt
@@ -481,10 +481,10 @@ PARAMS_SET
 ----------
 
 PARAMS_SET request modifies the settings retrieved by PARAMS_GET, i.e. it
-replaces ETHTOOL_SCOALESCE and ETHTOOL_SRINGPARAM ioctl commands. For each of
-these, relevant data attributes are contained in a corresponding nest
-attribute. Some of the attributes provided by PARAMS_GET are read only and
-cannot be set by PARAMS_SET request.
+replaces ETHTOOL_SCOALESCE, ETHTOOL_SRINGPARAM and ETHTOOL_SPAUSEPARAM ioctl
+commands. For each of these, relevant data attributes are contained in
+a corresponding nest attribute. Some of the attributes provided by
+PARAMS_GET are read only and cannot be set by PARAMS_SET request.
 
     ETHTOOL_A_PARAMS_COALESCE		(nested)	coalescing parameters
         ETHTOOL_A_COALESCE_RX_USECS	    (u32)
@@ -514,6 +514,10 @@ cannot be set by PARAMS_SET request.
         ETHTOOL_A_RING_RX_MINI_PENDING	    (u32)
         ETHTOOL_A_RING_RX_JUMBO_PENDING	    (u32)
         ETHTOOL_A_RING_TX_PENDING	    (u32)
+    ETHTOOL_A_PARAMS_PAUSE		(nested)	pause parameters
+        ETHTOOL_A_PAUSE_AUTONEG		    (bool)
+        ETHTOOL_A_PAUSE_RX		    (bool)
+        ETHTOOL_A_PAUSE_TX		    (bool)
 
 
 Request translation
@@ -542,7 +546,7 @@ ETHTOOL_SCOALESCE		ETHTOOL_MSG_PARAMS_SET
 ETHTOOL_GRINGPARAM		ETHTOOL_MSG_PARAMS_GET
 ETHTOOL_SRINGPARAM		ETHTOOL_MSG_PARAMS_SET
 ETHTOOL_GPAUSEPARAM		ETHTOOL_MSG_PARAMS_GET
-ETHTOOL_SPAUSEPARAM		n/a
+ETHTOOL_SPAUSEPARAM		ETHTOOL_MSG_PARAMS_SET
 ETHTOOL_GRXCSUM			ETHTOOL_MSG_SETTINGS_GET
 ETHTOOL_SRXCSUM			ETHTOOL_MSG_SETTINGS_SET
 ETHTOOL_GTXCSUM			ETHTOOL_MSG_SETTINGS_GET
diff --git a/net/ethtool/params.c b/net/ethtool/params.c
index 5af5448364c5..2751c30f6b64 100644
--- a/net/ethtool/params.c
+++ b/net/ethtool/params.c
@@ -306,7 +306,7 @@ static const struct nla_policy params_set_policy[ETHTOOL_A_PARAMS_MAX + 1] = {
 	[ETHTOOL_A_PARAMS_COMPACT]		= { .type = NLA_FLAG },
 	[ETHTOOL_A_PARAMS_COALESCE]		= { .type = NLA_NESTED },
 	[ETHTOOL_A_PARAMS_RING]			= { .type = NLA_NESTED },
-	[ETHTOOL_A_PARAMS_PAUSE]		= { .type = NLA_REJECT },
+	[ETHTOOL_A_PARAMS_PAUSE]		= { .type = NLA_NESTED },
 };
 
 static const struct nla_policy coalesce_policy[ETHTOOL_A_COALESCE_MAX + 1] = {
@@ -512,6 +512,54 @@ static int params_update_ring(struct genl_info *info, struct net_device *dev,
 	return (ret < 0) ? ret : 1;
 }
 
+static const struct nla_policy pause_policy[ETHTOOL_A_PAUSE_MAX + 1] = {
+	[ETHTOOL_A_PAUSE_UNSPEC]	= { .type = NLA_REJECT },
+	[ETHTOOL_A_PAUSE_AUTONEG]	= { .type = NLA_U8 },
+	[ETHTOOL_A_PAUSE_RX]		= { .type = NLA_U8 },
+	[ETHTOOL_A_PAUSE_TX]		= { .type = NLA_U8 },
+};
+
+/**
+ * params_update_pause() - update pause parameters from nested attribute
+ * @info: genl_info pointer (only needed for extack)
+ * @dev:  device to set pause parameters for
+ * @nest: ETHTOOL_A_PARAMS_PAUSE nested attribute
+ *
+ * Return: negative code on error, 0 if no change, >0 if there was a change
+ */
+static int params_update_pause(struct genl_info *info, struct net_device *dev,
+			struct nlattr *nest)
+{
+	struct nlattr *tb[ETHTOOL_A_RING_MAX + 1];
+	struct ethtool_pauseparam data = {};
+	bool mod = false;
+	int ret;
+
+	if (!nest)
+		return 0;
+	if (!dev->ethtool_ops->get_pauseparam ||
+	    !dev->ethtool_ops->set_pauseparam)
+		return -EOPNOTSUPP;
+	dev->ethtool_ops->get_pauseparam(dev, &data);
+
+	ret = nla_parse_nested(tb, ETHTOOL_A_PAUSE_MAX, nest, pause_policy,
+			       info->extack);
+	if (ret < 0)
+		return ret;
+
+	if (ethnl_update_u32(&data.autoneg, tb[ETHTOOL_A_PAUSE_AUTONEG]))
+		mod = true;
+	if (ethnl_update_u32(&data.rx_pause, tb[ETHTOOL_A_PAUSE_RX]))
+		mod = true;
+	if (ethnl_update_u32(&data.tx_pause, tb[ETHTOOL_A_PAUSE_TX]))
+		mod = true;
+
+	if (!mod)
+		return 0;
+	ret = dev->ethtool_ops->set_pauseparam(dev, &data);
+	return (ret < 0) ? ret : 1;
+}
+
 int ethnl_set_params(struct sk_buff *skb, struct genl_info *info)
 {
 	struct nlattr *tb[ETHTOOL_A_PARAMS_MAX + 1];
@@ -541,6 +589,11 @@ int ethnl_set_params(struct sk_buff *skb, struct genl_info *info)
 		goto out_ops;
 	if (ret)
 		req_mask |= ETHTOOL_IM_PARAMS_RING;
+	ret = params_update_pause(info, dev, tb[ETHTOOL_A_PARAMS_PAUSE]);
+	if (ret < 0)
+		goto out_ops;
+	if (ret)
+		req_mask |= ETHTOOL_IM_PARAMS_PAUSE;
 
 	ret = 0;
 out_ops:
-- 
2.22.0

